soundwire: add higher bus clock rate support

bardliao opened the pull request:
Refactor SoundWire clock support. And add multi-lane support.

bardliao committed:
soundwire: generic_bandwidth_allocation: add lane in sdw_group_params

All active streams with the same parameters are grouped together and the
params are stored in the sdw_group struct. We compute the required
bandwidth for each group. However, each lane has individual bandwidth.
Therefore, we should separate different lanes in different params groups.
Add lane variable to separate params groups.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart commented:
if you select the 9.6 MHz clock, you also need to change the default_col_size to 8 instead of 4. Otherwise the frame size will not be correct.

bardliao committed:
Soundwire: generic_bandwidth_allocation: set frame shape on fly

We need to recalculate frame shape when sdw bus clock is changed.
And need to make sure all Peripherals connected to the Manager support
dynamic clock change.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
Soundwire: add sdw_slave_get_scale_index helper

Currently, we only set peripheral frequency when the peripheral is
initialized. However, curr_dr_freq may change to get required bandwidth.
For example, curr_dr_freq may increase from 4.8MHz to 9.6MHz when the
4th stream is opened. Add a helper to get the scale index so that we can
get the scale index and program it.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
Soundwire: stream: program BUSCLOCK_SCALE

We need to program bus clock scale to adjust the bus clock if current
bus clock doesn't fit the bandwidth.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart commented on drivers/soundwire/stream.c:
```diff
@@ -637,8 +637,32 @@ static int sdw_notify_config(struct sdw_master_runtime *m_rt)
+			addr1 = SDW_SCP_BUSCLOCK_SCALE_B0;
+		}
+		/* Program SDW_SCP_BUSCLOCK_SCALE */
+		ret = sdw_write_no_pm(slave, addr1, scale_index);
```
careful here, not all peripherals support this register, and even worse not all peripherals support a dynamic clock scaling while audio is streaming.

so I think you need to make sure all devices are SDCA, we track this property somewhere.

And depending on the clock stale the frame size also needs to change accordingly. I don't see where the frame size is selected.

bardliao replied:
For those don't support dynamic clock scaling, I guess we should always use the max clock, right?

plbossart replied:
I think you should use the lower clock always, unless all components are SCDA.

plbossart replied:
The main issue is that the default configuration for LNL+ supports 2 frequencies, but the RVPs and AIOC are all with SDCA devices.
The problem starts if someone starts using non-SDCA devices with LNL+, then we might change the clock. So we've got to guard against this.

For older platforms, there is only one frequency anyways so there's no choice.

plbossart replied:
One easy solution is to always ignore default_col from ACPI and generate it at run time in the kernel. The default rows is correct.

bardliao replied:
> I think you should use the lower clock always, unless all components are SCDA.

But then we may meet the insufficient bandwidth issue. Why not use the highest clock always? It may waste power consumption, but at least we can fully use the bandwidth capability.

plbossart replied:
The higher clock is problematic on some designs, you may run into electrical/timing issues. So my recommendation is to ONLY use the higher clock starting with LNL and all SDCA setups.

bardliao replied:
> careful here, not all peripherals support this register, and even worse not all peripherals support a dynamic clock scaling while audio is streaming.
> 
> so I think you need to make sure all devices are SDCA, we track this property somewhere.
> 

I check if sdca version > 0. Not sure what SDCA version supports. It seems to be SDCA 1.0, but rt722 which complies SDCA 0.8rc1 supports this.
```
			if (s_rt->slave->sdca_data.interface_revision == 0) {
				dev_err(&s_rt->slave->dev, "The Peripheral doesn't comply with SDCA\n");
				return -EINVAL;
			}
```

plbossart replied:
I think all you need is 
````c
	if (!slave->id.class_id)
````

presumably all devices which set this bit in the ADR support the scaling registers, but better check with Cirrus Logic and Realtek

bardliao commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -534,14 +535,26 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+	if (curr_dr_freq == 19200000)
+		mstr_prop->default_col = 8;
+	else
+		mstr_prop->default_col = 4;
```
@plbossart Do you have a rule of selecting frame shape? Currently, I use a hack to test.

plbossart replied:
I think you need to start from the root frequency.
19.2: default_row = 50
12: default_row = 125
12.288: default_row = 64

Then you need to take the bus frequency, multiply it by to to get actual DDR rate, divide it by the frame rate and again by default row and you get the default_col.

bardliao replied:
> I think you need to start from the root frequency. 19.2: default_row = 50 12: default_row = 125 12.288: default_row = 64
> 
> Then you need to take the bus frequency, multiply it by to to get actual DDR rate, divide it by the frame rate and again by default row and you get the default_col.

Changed to `default_col = curr_dr_freq / m_rt->stream->params.rate / mstr_prop->default_row;`

plbossart commented on drivers/soundwire/stream.c:
```diff
@@ -1693,6 +1692,7 @@ static int _sdw_deprepare_stream(struct sdw_stream_runtime *stream)
 				return ret;
 			}
 		}
+		bus->params.bandwidth -= bandwidth - multi_lane_bandwidth;
```
not able to follow the commit message:
a) typo on "the bandwidth will of the stream" makes it unclear
b) "when compute_params() is called in sdw_prepare_stream()" The function modified is sdw_deprepare_stream, not sure why there's a reference to the prepare step.
c) it's not clear if the code is just a move or if bus->params.bandwidth can change inside of bus->compute_params().

bardliao replied:
The issue is that I will get "Compute transport params failed: -22" error without this commit. The issue is that we count bandwidth of all m_rt including the stream that is depreparing in bus->compute_params()->sdw_compute_port_params(). If we subtract bandwidth before calling bus->compute_params(), sdw_compute_bus_params() will select the clock that doesn't count the bandwidth of the stream that is depreparing.
To answer your questions.
a) That should be "the bandwidth of the stream"
b) Sorry, that was a typo, too. It should be "when compute_params() is called in sdw_deprepare_stream()"
c) bus->params.bandwidth will not be changed before multilane support added. However, bus->params.bandwidth will be modified when multilane is used.

plbossart commented on drivers/soundwire/stream.c:
```diff
@@ -1665,12 +1668,24 @@ static int _sdw_deprepare_stream(struct sdw_stream_runtime *stream)
-		bus->params.bandwidth -= m_rt->stream->params.rate *
-			m_rt->ch_count * m_rt->stream->params.bps;
+		bandwidth = m_rt->stream->params.rate * m_rt->ch_count * m_rt->stream->params.bps;
+		bus->params.bandwidth -= bandwidth - multi_lane_bandwidth;
```
this should be fixed here instead of having a separate patch "soundwire: subtract bandwidth after compute_params in deprepare_stream" which has problems too.

bardliao replied:
Don't you think they are 2 different issues? The needed of moving the bus->params.bandwidth change after bus->compute_params() is not because we added multi-lane support. The issue exists today. We didn't meet the issue because we use a fixed clock rate. The issue only happens when the 4th stream is closed. The scenario is that we use a higher clock and new frame shape when the 4th stream is opened. And when any stream is closed, i.e. the active stream number become 3, sdw_compute_bus_params() will select lower clock rate and the original frame shape. But, sdw_compute_group_params() still calculate the `payload_bw` for 4 streams since the m_rt is not removed yet. Then sdw_compute_group_params() will return -EINVAL because `column_needed > sel_col - 1`.

plbossart replied:
sure, then move this patch earlier

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -506,6 +508,12 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+					__func__, curr_dr_freq);
+				i++;
+				goto select_clk;
+			}
```
I am not sure why you would only increase the clock when there is a single lane. The clock should be increased even in the multi-lane case when there's no bandwidth left?

I must be missing something.

bardliao replied:
That is about the priority of using multi-lane or higher clock rate. At this point, we tried multi-lane but can't find an available lane and it will return -EINVAL if it reaches the highest clock rate. But yeah, I didn't consider about using multi-lane with higher clock rate. I assume that we always have enough bandwidth if we can find an extra lane. Let me redo this.

plbossart replied:
yes, lanes should be used before cranking the speed up, but an error should only be thrown when multilane+higher speed combined doesn't support the required bandwidth.

plbossart commented on drivers/soundwire/bus.c:
```diff
@@ -1354,6 +1343,33 @@ static int sdw_slave_set_frequency(struct sdw_slave *slave)
+	 * frequency base and scale registers are required for SDCA
+	 * devices. They may also be used for 1.2+/non-SDCA devices.
+	 * Driver can set the property, we will need a DisCo property
+	 * to discover this case from platform firmware.
```
not able to follow this comment, consider rewording.

bardliao replied:
It was just copy and paste from the existing code.

plbossart replied:
maybe clarify the comment 

"Driver can set the property directly, for now there's no DisCo property to discover support for the scaling registers from platform firmware'

bardliao committed:
soundwire: generic_bandwidth_allocation: check required freq accurately

Currently, we check curr_dr_freq roughly by "if (curr_dr_freq <=
bus->params.bandwidth)" in sdw_compute_bus_params() and check it
accurately in sdw_select_row_col(). It works if we only support one
freq. But, we need to check it accurately in sdw_select_row_col() to
give it a chance to use a higher freq or use multi-lane.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
Revert "soundwire: generic_bandwidth_allocation: add lane in sdw_group_params"

This reverts commit 72e43ef1bccb9302ca22ee1aa583789f3187dd74.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
Revert "soundwire: generic_bandwidth_allocation: select data lane"

This reverts commit e56ee9f967eded851b60c6077374428f68979d6f.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao commented:
On the second thought, I think the existing code partially support dynamic clock change already. What missed is change frame shape dynamically. So, it will be better if we add the dynamic frame shape support commit first and multi-lane support commits late. It is more reviewable, and the logic is clearer.

plbossart commented on drivers/soundwire/bus.c:
```diff
@@ -1276,23 +1276,12 @@ int sdw_configure_dpn_intr(struct sdw_slave *slave,
 }
 
-static int sdw_slave_set_frequency(struct sdw_slave *slave)
+u8 sdw_slave_get_scale_index(struct sdw_slave *slave, u8 *base)
```
commit message:

when the 4th stream is open -> opened?
update the sacle index if needed. -> sample?

plbossart commented on drivers/soundwire/stream.c:
```diff
@@ -637,8 +637,32 @@ static int sdw_notify_config(struct sdw_master_runtime *m_rt)
+		}
+	}
+
+
```
spurious line

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -377,6 +378,18 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+		/* Check if default_col can be changed */
+		list_for_each_entry(s_rt, &m_rt->slave_rt_list, m_rt_node) {
+			if (!s_rt->slave->id.class_id) {
+				dev_err(&s_rt->slave->dev, "The Peripheral doesn't comply with SDCA\n");
```
you probably need a comment to explain that some devices expose the class ID but can't support dynamic scaling.

I would actually move this loop to a helper that  can be modified/quirked if needed, e.g. is_clock_scaling_supported().

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -360,7 +361,11 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
 
-		if (curr_dr_freq <= bus->params.bandwidth)
+		frame_int = mstr_prop->default_row * mstr_prop->default_col;
+		frame_freq = curr_dr_freq / frame_int;
```
not following what 'frame_int' refers to? Frame_size or frame_bitslots maybe?

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -360,7 +361,11 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+		frame_int = mstr_prop->default_row * mstr_prop->default_col;
+		frame_freq = curr_dr_freq / frame_int;
+
+		if ((curr_dr_freq - (frame_freq * SDW_FRAME_CTRL_BITS)) <
```
I think this is partially wrong, because we never use column0 for audio. so for a 50 row frame, this calculation would be off by 2 bits times the frame_freq. I think you need to use curr_dr_freq - default_row * frame_freq.

bardliao replied:
This is copy from `sdw_select_row_col()`. Do we need to do the same change in `sdw_select_row_col()`, too?

plbossart replied:
yeah, I think so. I don't think anyone can use column0 with 19.2 or 12.228 root frequencies. It's only possible for the 125-row frame but that's such a corner case.

But the formula could be simplifed to avoid a division btw:

curr_dr_freq - (curr_dr_freq / (sdw_rows[r] * sdw_cols[c]) * sdw_rows[r])

curr_d_freq - curr_dr_freq / sdw_cols[c] < bandwidth
curr_d_freq * (sdw_cols[c] - 1) < bandwidth * sdw_cols[c]

plbossart commented on drivers/soundwire/stream.c:
```diff
@@ -1693,16 +1693,6 @@ static int _sdw_deprepare_stream(struct sdw_stream_runtime *stream)
-				return ret;
-			}
-		}
-
```
The clocks are always updated on prepare/deprepare, I am not sure why you want to remove this.

bardliao replied:
I am not sure if we really need to update clocks on both prepare/deprepare. We update clock in sdw_compute_bus_params(), but we still reserve the bit slots for the deprepared stream in sdw_compute_port_params(). That doesn't look correct to me.
This is to see if any issue will happen if I remove this. Looks like CI has told me that compute_params() can not be removed.

plbossart replied:
let me re-explain. We have two bank switches, one on prepare and one on trigger. This is intentional, we want to go to the 'new' frequency and let the clock toggle for some time before enabling all the channels on the trigger.

So the point is that we do need to update the parameters and reduce the clock on deprepare. I am not sure why the prepare and de-prepare parts would become asymmetrical?

bardliao replied:
> let me re-explain. We have two bank switches, one on prepare and one on trigger. This is intentional, we want to go to the 'new' frequency and let the clock toggle for some time before enabling all the channels on the trigger.
> 
> So the point is that we do need to update the parameters and reduce the clock on deprepare.

Agree, but m_rt is removed after deprepared and list_for_each_entry(m_rt, &bus->m_rt_list, bus_node) is used by _sdw_compute_port_params(), sdw_compute_group_params(), etc. As a result, the bit slots of the deprepared stream are still reserved. If we reduce the clock, we will get a no sufficient bandwidth error.

> I am not sure why the prepare and de-prepare parts would become asymmetrical?

I don't think it is asymmetrical. Below is the sequence when we start and stop aplay.
intel_hw_params
 -> sdw_stream_add_master
asoc_sdw_prepare
 -> sdw_prepare_stream

asoc_sdw_hw_free
 -> sdw_deprepare_stream
intel_hw_free
 -> sdw_stream_remove_master

plbossart replied:
Not following this. Do we agree when the stream is deprepared, it should use exactly zero bitslots? I don't understand why the bits would remain reserved.

bardliao replied:
> Not following this. Do we agree when the stream is deprepared, it should use exactly zero bitslots? 

Yes, totally agree.

> I don't understand why the bits would remain reserved.

That is the issue in current code and that is because the m_rt is not removed at that point.

plbossart replied:
but the code above still accounts for the bandwidth of the stream, so the stream is still taken into account. 

It makes no sense to me to reduce the bandwidth count *after* performing the bandwidth allocation.

Sorry, completely lost here.

bardliao replied:
Moving the bandwidth reduction count after performing the bandwidth allocation is based on the fact that the m_rt has not been removed when we allocate the bit slots. To fix the issue, I can either move the bandwidth reduction count after performing the bandwidth as the current PR does or remove the m_rt before allocating the bit slots.
I choose the former because removing m_rt first seems break the add m_rt -> prepare -> deprepare ->remove m_rt sequence.
I know the side effects of moving the bandwidth reduction count after performing the bandwidth allocation is that we may keep the sdw bus clock or lane that is not needed. But it will happen only when we reach the threshold that need higher clock rate or more lane. So, I think it is acceptable.

plbossart replied:
That's a bit convoluted really.

The bits for the stream should no longer be accounted for after this routine:

````c
int asoc_sdw_hw_free(struct snd_pcm_substream *substream)
{
	struct snd_soc_pcm_runtime *rtd = snd_soc_substream_to_rtd(substream);
	struct sdw_stream_runtime *sdw_stream;
	struct snd_soc_dai *dai;

	/* Find stream from first CPU DAI */
	dai = snd_soc_rtd_to_cpu(rtd, 0);

	sdw_stream = snd_soc_dai_get_stream(dai, substream->stream);
	if (IS_ERR(sdw_stream)) {
		dev_err(rtd->dev, "no stream found for DAI %s\n", dai->name);
		return PTR_ERR(sdw_stream);
	}

	return sdw_deprepare_stream(sdw_stream);
}
````

but the problem is that we have a half-baked dai handling and comments

````c
static int
intel_hw_free(struct snd_pcm_substream *substream, struct snd_soc_dai *dai)
{
	struct sdw_cdns *cdns = snd_soc_dai_get_drvdata(dai);
	struct sdw_intel *sdw = cdns_to_intel(cdns);
	struct sdw_cdns_dai_runtime *dai_runtime;
	int ret;

	dai_runtime = cdns->dai_runtime_array[dai->id];
	if (!dai_runtime)
		return -EIO;

	/*
	 * The sdw stream state will transition to RELEASED when stream->
	 * master_list is empty. So the stream state will transition to
	 * DEPREPARED for the first cpu-dai and to RELEASED for the last
	 * cpu-dai.
	 */
	ret = sdw_stream_remove_master(&cdns->bus, dai_runtime->stream);
	if (ret < 0) {
		dev_err(dai->dev, "remove master from stream %s failed: %d\n",
			dai_runtime->stream->name, ret);
		return ret;
	}
````

That comment is wrong, the stream state will transition to DEPREPARED in the dailink hw_free(), and then to RELEASED when the last DAI hw_free() is called

That still doesn't really tell me why the bits are not released during the sdw_deprepare_stream() part. There should be no dependency on the m_rt list.

bardliao replied:
asoc_sdw_hw_free() is called before intel_hw_free(). IOW, sdw_deprepare_stream() is called before sdw_stream_remove_master(). So, m_rt is not removed yet when sdw_deprepare_stream() is called.

_sdw_deprepare_stream
-> bus->compute_params(bus)
-> sdw_compute_params()
    -> sdw_compute_bus_params(bus)
    -> sdw_compute_port_params(bus)
````c
                        port_bo = 1;

                        list_for_each_entry(m_rt, &bus->m_rt_list, bus_node) {
                                sdw_compute_master_ports(m_rt, &params[i], &port_bo, hstop);
                        }

                        hstop = hstop - params[i].hwidth;

````
And we calculate hstart, hstop, and offset in sdw_compute_master_ports()
````c
                t_data.hstart = hstart;
                t_data.hstop = hstop;
                t_data.block_offset = *port_bo;
                t_data.sub_block_offset = 0;
                (*port_bo) += bps * ch;

````
So that we allocate bit slots for each m_rt.

plbossart replied:
yeah, that's clearly wrong, the generic bit allocation should take the states into account, and skip the deprepared streams.

See https://github.com/thesofproject/linux/pull/5177 for compile-tested-only proposal.

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -353,6 +391,10 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
 	}
 
+	m_rt = list_last_entry(&bus->m_rt_list, struct sdw_master_runtime, bus_node);
+	s_rt = list_first_entry(&m_rt->slave_rt_list, struct sdw_slave_runtime, m_rt_node);
```
not following what this is supposed to do and why you are not dealing with M and S in symmetrical ways, the last/first isn't clear to me at all.

bardliao replied:
I use the last m_rt is because that one is recently added to the bus and it is the m_rt that we need to handle. And for the s_rt, it doesn't really matter to use what s_rt in the m_rt. The s_rt will be used to find the peripheral lane, and we will check if the lane is connected to all peripherals.

plbossart replied:
I don't think that's correct @bardliao. When there is a new stream, we need to compute the parameters for all streams and all _rt contexts. Think of it this way: at any bank switch the position of all bits could change. It could very well be that the new stream will have higher priority in the bandwidth allocation than the new one.

IOW, the generic allocation is NOT a purely additive mechanism. At every bank switch, the bus allocation is a blank slate and ALL streams/ports can be re-programmed.

bardliao replied:
Changed to `list_for_each_entry(m_rt, &bus->m_rt_list, bus_node)` to check all m_rt

bardliao committed:
soundwire: generic_bandwidth_allocation: correct clk_freq check in sdw_select_row_col

The bits in Column 0 of Rows 0 to 47 are for control word and cannot be
used for audio. In practice, entire Column 0 is skipped.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
soundwire: stream: set DEPREPARED state earlier

The existing logic is problematic in that we deprepare all the ports,
but still take into account the stream for bit allocation by just
walking through the bus->m_rt list.

This patch sets the state earlier, so that such DEPREPARED streams can
be skipped in the bandwidth allocation (to be implemented in a
follow-up patch).

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart committed:
soundwire: generic_bandwidth_allocation: skip DEPREPARED streams

We should not blindly walk through all the m_rt list, since it will
have the side effect of accounting for deprepared streams.

This behavior is the result of the split implementation where the
dailink hw_free() handles the stream state change and the bit
allocation, and the dai hw_free() modifies the m_rt list. The bit
allocation ends-up using m_rt entries in zombie state, no longer
relevant but still used.

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.dev>
Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart commented on drivers/soundwire/stream.c:
```diff
@@ -637,8 +637,37 @@ static int sdw_notify_config(struct sdw_master_runtime *m_rt)
+			dev_dbg(&slave->dev, "%s: The Peripheral doesn't comply with SDCA\n",
+				__func__);
+			break;
+		}
```
nit-pick: I would make this a helper that can be quirked if needed, e.g. in a case with CS42l42 with multiple frequencies. The class_id may not be enough in all cases.

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -329,6 +329,27 @@ static int sdw_select_row_col(struct sdw_bus *bus, int clk_freq)
+	}
+	return true;
+}
+
```
humm, should this helper be introduced in the previous patch? see other comment.

plbossart replied:
comment still applies

bardliao replied:
It is a bit different between here and the previous commit. We are in the list_for_each_entry(slave, &bus->slaves, node) loop in previous commit, so the helper we need in the previous commit is to check if a single Peripheral supports dynamic scaling. But here the helper checks all Peripherals in a bus. So, I use different helpers in the 2 commits.

plbossart replied:
right right, I meant something like:

````c
	list_for_each_entry(m_rt, &bus->m_rt_list, bus_node) {
		list_for_each_entry(s_rt, &m_rt->slave_rt_list, m_rt_node) {
                       if (!is_clock_scaling_supported_by_slave(s_rt->slave)
````

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -380,6 +406,10 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
 
+	/* Get the last m_rt that is recently added to the bus. */
+	m_rt = list_last_entry(&bus->m_rt_list, struct sdw_master_runtime, bus_node);
+	mstr_prop->default_col = curr_dr_freq / m_rt->stream->params.rate / mstr_prop->default_row;
```
oh no no no, **big confusion here**!

The frame shape need to be computed with the SoundWire frame rate, which can be different to all the stream rates. Think e.g. of the example I presented with a 96 kHz frame rate to support two streams, one at 48kHz and one at 48 kHz.

bardliao replied:
Right, right, fixed now.

bardliao committed:
soundwire: generic_bandwidth_allocation: select data lane

If a peripheral supports multi-lane, we can use data lane x to extend
the bandwidth. The patch suggests to select data lane x where x > 0
when bandwidth is not enough on data lane 0.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -354,6 +375,10 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
 		clk_buf = NULL;
 	}
 
+	/* If dynamic scaling is supported, don't try higher freq */
```
logical inversion in comment, should be "NOT supported"?

bardliao replied:
Fixed.

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -384,10 +422,56 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+			}
+		} else {
+			break;
+		}
```
I would revert the logic to do the break first and then you can make the code more readable with one level of indentation removed.

bardliao replied:
Good suggestion. Done.

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -133,94 +140,126 @@ static void sdw_compute_master_ports(struct sdw_master_runtime *m_rt,
 	/* Run loop for all groups to compute transport parameters */
-	for (i = 0; i < count; i++) {
-		port_bo = 1;
+	for (l = 0; l < SDW_MAX_LANES; l++) {
```
wondering if we can optimize and only deal with existing lanes. 90% of the time there will be a single lane, it's a bit odd to loop on all possible lanes.

plbossart reviewed (commented):
looks mostly good, I still think the code can be made a bit clearer/compact, see suggestions below.

plbossart commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -329,6 +329,27 @@ static int sdw_select_row_col(struct sdw_bus *bus, int clk_freq)
 	return -EINVAL;
 }
 
+static bool is_clock_scaling_supported(struct sdw_bus *bus)
```
commit message typo shpport -> support

plbossart commented on drivers/soundwire/stream.c:
```diff
@@ -627,6 +627,16 @@ static int sdw_notify_config(struct sdw_master_runtime *m_rt)
+	return slave->id.class_id;
+}
+EXPORT_SYMBOL(is_clock_scaling_supported_by_slave);
+
```
nit-pick: if this is exported, then maybe it should be in slave.c or or bus.c, this isn't really stream-related.

bardliao replied:
> nit-pick: if this is exported, then maybe it should be in slave.c or or bus.c, this isn't really stream-related.

Moved to bus.c

bardliao commented:
SOFCI TEST

bardliao commented:
SOFCI TEST

ujfalusi commented on drivers/soundwire/stream.c:
```diff
@@ -637,8 +637,37 @@ static int sdw_notify_config(struct sdw_master_runtime *m_rt)
+
+		if (!is_clock_scaling_supported_by_slave(slave)) {
+			dev_dbg(&slave->dev, "%s: The Peripheral doesn't comply with SDCA\n",
+				__func__);
```
No need to print the function name for dev_dbg()

bardliao replied:
removed

ujfalusi commented on drivers/soundwire/stream.c:
```diff
@@ -637,8 +637,37 @@ static int sdw_notify_config(struct sdw_master_runtime *m_rt)
+		if (!is_clock_scaling_supported_by_slave(slave)) {
+			dev_dbg(&slave->dev, "%s: The Peripheral doesn't comply with SDCA\n",
+				__func__);
+			break;
```
We `break` and not `continue` iow, if any of the codecs is not SDCA compliant, the scaling is not possible?
A comment would be nice.

and if it is not the first in the slaves we already configured previous codecs's SDW_SCP_BUSCLOCK_SCALE ?

bardliao replied:
> We `break` and not `continue` iow, if any of the codecs is not SDCA compliant, the scaling is not possible? A comment would be nice.

Yes, ALL Peripherals should be SDCA compliant.

> 
> and if it is not the first in the slaves we already configured previous codecs's SDW_SCP_BUSCLOCK_SCALE ?

Indeed, it will be an issue. I will use another loop to program SDW_SCP_BUSCLOCK_SCALE, thanks.

ujfalusi commented on drivers/soundwire/stream.c:
```diff
@@ -637,8 +637,37 @@ static int sdw_notify_config(struct sdw_master_runtime *m_rt)
+		/* Program SDW_SCP_BUSCLOCK_SCALE */
+		ret = sdw_write_no_pm(slave, addr1, scale_index);
+		if (ret < 0) {
+			dev_err(bus->dev, "SDW_SCP_BUSCLOCK_SCALE register write failed\n");
```
Since we are in a loop, probably worth printing for _which_ slave/address the write failed?

bardliao replied:
Yes, changed to &slave->dev

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -380,6 +397,9 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
 	}
 
+	mstr_prop->default_col = curr_dr_freq / mstr_prop->default_frame_rate /
+					mstr_prop->default_row;
```
nitpick: I would align this with the curr_dr_freq.

bardliao replied:
done

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -338,18 +338,56 @@ static bool is_clock_scaling_supported(struct sdw_bus *bus)
+ * @m_rt: Manager runtime
+ * @lane: Lane number
+ */
+static bool check_all_peripherals_connected(struct sdw_master_runtime *m_rt, unsigned int lane)
```
the function name is really neutral, no way to tell what it is checking, can you clarify it a bit?

bardliao replied:
The function is checking if the manager lane connects to all peripherals. So that all peripherals can use the manager lane. i.e, if one peripheral doesn't connect to that lane, the lane can not be used. Update the function description.

ujfalusi replied:
but the code is checking if at least one peripheral is connected to the lane (which is passed as a parameter)
So it is like is_plane_used() / is_plane_have_connected_peripheral() type of check, no?

ujfalusi replied:
right, that is an interesting query to name for sure...
is_lane_connected_to_all_peripherals() ?

bardliao replied:
updated

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -375,11 +413,60 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+				if (!slave_prop->lane_maps[l])
+					continue;
+
+				dev_dbg(bus->dev, "%s: trying lane %d\n", __func__, l);
```
No need for __func__ in dev_dbg()

bardliao replied:
removed

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -375,11 +413,60 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+
+				dev_dbg(bus->dev, "%s: trying lane %d\n", __func__, l);
+				required_bandwidth = 0;
+				list_for_each_entry(m_p_rt, &m_rt->port_list, port_node) {
```
This is the fourth nested loop in this function... can we split?

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -375,11 +413,60 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+				list_for_each_entry(m_p_rt, &m_rt->port_list, port_node) {
+					required_bandwidth += m_rt->stream->params.rate *
+						hweight32(m_p_rt->ch_mask) *
+						m_rt->stream->params.bps;
```
nitpick: I would align these to m_rt

bardliao replied:
done

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -375,11 +413,60 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+						slave_prop->lane_maps[l])) {
+						dev_dbg(bus->dev,
+							"some Peripherals are not connected to %d\n",
+							slave_prop->lane_maps[l]);
```
what does this means?
Some is connected, some is not?

Confusing.

bardliao replied:
Both Managers and Peripherals can have multiple lanes. They could be connected by any lane. For example, Peripheral A could connect to the Manager with lane 0, 1 and 2. And Peripheral A could connect to the Manager with lane 0, 1 and 3. In that case, we can only use lane 1 as the second lane.

ujfalusi commented on drivers/soundwire/stream.c:
```diff
@@ -1703,9 +1706,22 @@ static int _sdw_deprepare_stream(struct sdw_stream_runtime *stream)
+				continue;
+
+			bandwidth = m_rt->stream->params.rate * hweight32(p_rt->ch_mask) *
+				m_rt->stream->params.bps;
```
nitpick: I would align this to m_rt

bardliao replied:
done

ujfalusi commented on drivers/soundwire/stream.c:
```diff
@@ -637,8 +637,40 @@ static int sdw_notify_config(struct sdw_master_runtime *m_rt)
+		u8 base;
+
+		scale_index = sdw_slave_get_scale_index(slave, &base);
+		if (scale_index < 0)
```
u32 can rarely be less than 0...

bardliao replied:
Yes, changed to int, thanks.

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -338,18 +338,103 @@ static bool is_clock_scaling_supported(struct sdw_bus *bus)
+			}
+		}
+		if (i == SDW_MAX_LANES) {
+			dev_dbg(&s_rt->slave->dev, "%d is not connected\n", lane);
```
lane is not connected to all peripherals, to be precise, since it can be connected to some, no?

bardliao replied:
Yes, it means the lane is not connected to the peripheral. Change it to `dev_dbg(&s_rt->slave->dev, "%d is not connected to %s\n", lane, dev_name(&s_rt->slave->dev));`

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -338,18 +338,103 @@ static bool is_clock_scaling_supported(struct sdw_bus *bus)
+			if (slave_prop->lane_maps[i] == lane) {
+				dev_dbg(&s_rt->slave->dev,
+					"lane_maps[%d] is connected to %d\n",
+					i, slave_prop->lane_maps[i]);
```
just to understand, this is trying to say that the peripheral's lane `i` is connected to manager's `lane`, right?

bardliao replied:
> just to understand, this is trying to say that the peripheral's lane `i` is connected to manager's `lane`, right?

Yes, lane_maps is the connection map. lane_maps[x] = y means peripheral's lane x is connected to manager lane y.

ujfalusi replied:
Can we print this in a human readable fashion, like `P lane %d is connected to M lane %d`  ?

bardliao replied:
Yes, done

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -338,18 +338,103 @@ static bool is_clock_scaling_supported(struct sdw_bus *bus)
+		if (!slave_prop->lane_maps[l])
+			continue;
+
+		dev_dbg(bus->dev, "trying lane %d\n", l);
```
is this only probe time or runtime?
Will it be noisy, can we get away w/o it?

bardliao replied:
> is this only probe time or runtime? Will it be noisy, can we get away w/o it?

This will be called at runtime. I will remove it.

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -338,18 +338,103 @@ static bool is_clock_scaling_supported(struct sdw_bus *bus)
+			}
+			m_lane = slave_prop->lane_maps[l];
+			dev_dbg(&s_rt->slave->dev,
+				"M lane %d P lane %d can be used\n",
```
How to read this print?
The given M lane is going to be connected to the given P lane, or something else?

bardliao replied:
The given M lane is connected to the given P lane, so this lane can be used.

ujfalusi replied:
Used for what?
In theory a peripheral can be connected to multiple manager lane?

Within is_lane_connected_to_all_peripherals() we have already printed all the connections with mentioning P and M lane with the peripheral's device used for dev_dbg(), this is going to state the same but just a bit differently, so perhaps:
`M lane %d is used' and that's it?

Or consolidate the prints somehow?

bardliao replied:
Yes, changed to "M lane %d is used"

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -338,18 +338,103 @@ static bool is_clock_scaling_supported(struct sdw_bus *bus)
+			 * Use non-zero manager lane, subtract the lane 0
+			 * bandwidth that is already calculated
+			 */
+			bus->params.bandwidth -= required_bandwidth;
```
this cannot underflow, right?

bardliao replied:
Right, it is unsigned int.

ujfalusi reviewed (commented):
@bardliao, nothing big stands out, just few smaller ones.

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -338,18 +338,103 @@ static bool is_clock_scaling_supported(struct sdw_bus *bus)
+			}
+		}
+		if (i == SDW_MAX_LANES) {
+			dev_dbg(&s_rt->slave->dev, "%d is not connected to %s\n",
```
Probably simplify it to: `M lane %d is not connected` (the dev_name(&s_rt->slave->dev) is printed by dev_dbg())

bardliao replied:
done

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -338,18 +338,103 @@ static bool is_clock_scaling_supported(struct sdw_bus *bus)
+			if (!is_lane_connected_to_all_peripherals(m_rt,
+				slave_prop->lane_maps[l])) {
+				dev_dbg(bus->dev,
+					"some Peripherals are not connected to %d\n",
```
`Not all Peripherals are not connected to M lane %d, M lane is not usable` ? If it is correct.

bardliao replied:
done

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -375,11 +460,27 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+			/*
+			 * Find available manager lanes that connected to the first Peripheral.
+			 * No need to check all Peripherals available lanes because we can't use
+			 * multi-lane if we can't find any available lane for the first Peripheral.
```
get_manager_lane() will check all peripherals lane connection via a call to is_lane_connected_to_all_peripherals().
Isn't this a bit contradicting to this comment?

bardliao replied:
That means something different. "No need to check all Peripherals" here means we only need to check the available lanes of the first Peripheral. Then, `is_lane_connected_to_all_peripherals()` will check if the lane can be used for all Peripherals. We don't need to repeat checking all Peripherals because one lane can not be used if the lane is not connected to the first Peripheral.

bardliao replied:
Reworded it. Hope it is more readable. :)

ujfalusi reviewed (commented):
@bardliao, I'm still wrapping my head around the debug prints, sorry.

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -338,18 +338,100 @@ static bool is_clock_scaling_supported(struct sdw_bus *bus)
+			if (!is_lane_connected_to_all_peripherals(m_rt,
+				slave_prop->lane_maps[l])) {
+				dev_dbg(bus->dev,
+					"Not all Peripherals are not connected to M lane %d, M lane is not usablen\n",
```
For the sake of avoiding long line, I think we should drop the `M lane is not usable`

bardliao replied:
Yes, done

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -338,18 +338,100 @@ static bool is_clock_scaling_supported(struct sdw_bus *bus)
+		for (i = 1; i < SDW_MAX_LANES; i++) {
+			if (slave_prop->lane_maps[i] == lane) {
+				dev_dbg(&s_rt->slave->dev,
+					"P lane %d is connected to M lane %d\n",
```
It might be better to swap the P and M lane in the sentence?
`M lane %d is connected to P lane %d`

bardliao replied:
updated

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -393,6 +490,30 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
+			slave_prop = &s_rt->slave->prop;
+			for (l = 1; l < SDW_MAX_LANES; l++) {
+				if (slave_prop->lane_maps[l] == m_lane) {
+					dev_dbg(&s_rt->slave->dev, "Set Peripheral lane = %d\n", l);
```
Should we stick with `P lane` for consistency over `Peripheral lane` print use only here?
`Set P lane %d for ports` perhaps, if select makes sense in this context. 

or do we even need to print anything in here?

bardliao replied:
updated

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -393,6 +490,30 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
 	}
+out:
+	/* multilane can be used */
+	if (m_lane > 0) {
```
I think you should initialize m_lane to 0 at declaration.

bardliao replied:
done, thanks.

ujfalusi commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -338,17 +338,99 @@ static bool is_clock_scaling_supported(struct sdw_bus *bus)
+			if (!is_lane_connected_to_all_peripherals(m_rt,
+				slave_prop->lane_maps[l])) {
+				dev_dbg(bus->dev,
+					"Not all Peripherals are not connected to M lane %d\n",
```
you have a leftover 'not' word:
"Not all Peripherals are **not** connected to M lane %d\n"

ujfalusi reviewed (commented):
@bardliao, only one typo that I can see, sorry.

ujfalusi reviewed (dismissed):
@bardliao, thank you for the updates, it looks good to me.

bardliao commented:
@ranj063 Do you have any comment on this PR?

ranj063 commented on drivers/soundwire/stream.c:
```diff
@@ -1653,15 +1653,24 @@ static int _sdw_deprepare_stream(struct sdw_stream_runtime *stream)
+	 * first mark the state as DEPREPARED so that it is not taken into account
+	 * for bit allocation
+	 */
+	state = stream->state;
```
this is already set during init

bardliao replied:
Removed, thanks.

ranj063 commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -240,6 +240,9 @@ static int sdw_get_group_count(struct sdw_bus *bus,
 		return -ENOMEM;
 
 	list_for_each_entry(m_rt, &bus->m_rt_list, bus_node) {
+		if (m_rt->stream->state == SDW_STREAM_DEPREPARED)
```
2 typos in the commit message, zoombie->zombie & not longer -> no longer

bardliao replied:
fixed

ranj063 commented on drivers/soundwire/stream.c:
```diff
@@ -637,8 +637,40 @@ static int sdw_notify_config(struct sdw_master_runtime *m_rt)
+	else
+		addr1 = SDW_SCP_BUSCLOCK_SCALE_B0;
+
+	/* Progrem SDW_SCP_BUSCLOCK_SCALE is all Peripherals comply with SDCA */
```
typos: Progrem -> program & is all -> if all

bardliao replied:
fixed

ranj063 commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -304,7 +304,6 @@ static int sdw_compute_port_params(struct sdw_bus *bus)
 static int sdw_select_row_col(struct sdw_bus *bus, int clk_freq)
 {
 	struct sdw_master_prop *prop = &bus->prop;
-	int frame_int, frame_freq;
```
typo: can not is one word

bardliao replied:
fixed

ranj063 commented on drivers/soundwire/bus.h:
```diff
@@ -151,6 +151,7 @@ struct sdw_transport_data {
 	int hstop;
 	int block_offset;
 	int sub_block_offset;
+	unsigned int lane;
```
typo in commit message: store ->stored

bardliao replied:
fixed

lgirdwood commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -380,6 +397,9 @@ static int sdw_compute_bus_params(struct sdw_bus *bus)
 		return -EINVAL;
 	}
 
+	mstr_prop->default_col = curr_dr_freq / mstr_prop->default_frame_rate /
+				 mstr_prop->default_row;
+
```
Do we need to validate any of these are non 0 ? i.e. if they come from external ACPI or DISCO data ?

bardliao replied:
> Do we need to validate any of these are non 0 ? i.e. if they come from external ACPI or DISCO data ?

In theory they should not be 0. SDW bus will not work if they are 0. But yeah, it is worthy to check it.

lgirdwood commented on drivers/soundwire/generic_bandwidth_allocation.c:
```diff
@@ -287,7 +343,7 @@ static int sdw_compute_port_params(struct sdw_bus *bus)
 	/* Compute transport parameters for grouped streams */
 	ret = sdw_compute_group_params(bus, params,
-				       &group.rates[0], group.count);
+				       &group.rates[0], &group.lanes[0], group.count);
```
nitpick: could we just pass ```group``` to this API ?

bardliao replied:
> nitpick: could we just pass `group` to this API ?

Yes, updated

lgirdwood reviewed (commented):
LGTM, just minor questions.

ranj063 reviewed (commented):
@bardliao could you please also add your sign-off on Pierre's patches?

bardliao commented:
> @bardliao could you please also add your sign-off on Pierre's patches?

Yes, added.
